Adds error handling on receive and create

LukeHankey opened the pull request:
# Fixes:
- [x] A `TypeError` is raised when something other than JSON is sent to EventRequest.
- [x] `pydantic.error_wrappers.ValidationError` is raised when data fields are not found.
- [x] A `KeyError` is raised when the keys are not found in data for the respective EventData type.

# Features:
- [x] A new error should be raised when creating a game with an already existing code: `RoomAlreadyExistsError`

LukeHankey committed:
fix: TypeError was raised if an incorrect data type (non-json) was sent

kappq reviewed (changes requested):
This doesn't seem to catch any errors. I've tried:
- Sending invalid JSON (e.g. a comma instead of a colon)
- Send JSON with a missing field (e.g. missing the "type" field)
- Send JSON that doesn't pass the validators (e.g. sending a connect request with the connection type "create" that has the "difficulty" field missing)

Anyway, we probably want some kind of feedback when invalid JSON is sent, rather than just returning from the function.

LukeHankey commented:
> This doesn't seem to catch any errors. I've tried:
> 
> * Sending invalid JSON (e.g. a comma instead of a colon)
> * Send JSON with a missing field (e.g. missing the "type" field)
> * Send JSON that doesn't pass the validators (e.g. sending a connect request with the connection type "create" that has the "difficulty" field missing)
> 
> Anyway, we probably want some kind of feedback when invalid JSON is sent, rather than just returning from the function.

It's for **non-json** that is sent. It's unlikely, but better to catch the error. I agree with sending some response back instead of just disconnecting.

LukeHankey committed:
fix: Adds error handling to client.receive() for incorrect input types

New error codes (4002, 4003, 4004) to be sent back to the client and then disconnected.
Frontend will have to check if any of these error codes are sent back during a live game by checking that the event type is not connect

LukeHankey committed:
feat: Add new error: RoomAlreadyExistsError

LukeHankey committed:
Adds error handling for new error and corrects types

LukeHankey committed:
Adds JSONDecodeError if no data is sent at all

LukeHankey committed:
Remove VALIDATION_ERROR code as not used

LukeHankey committed:
removes unnecessary await/async from ConnectionManager.connect

LukeHankey committed:
Merge branch 'main' into incorrect-data

kappq commented on server/codes.py:
```diff
@@ -5,4 +5,11 @@ class StatusCode(IntEnum):
     """Custom status codes for a WebSocket event."""
 
     SUCCESS = 4000
+
```
I don't know if there should be a new line between enum variants.

kappq commented on server/main.py:
```diff
@@ -238,19 +261,19 @@ async def room(websocket: WebSocket) -> None:
         while True:
             event = await client.receive()
+            if event is None:
+                return
```
We might want to continue if the event is `None`.

LukeHankey replied:
In what case? If event is None, then an error was raised

kappq replied:
> In what case? If event is None, then an error was raised

Yeah, but if, for example, the data was invalid, we don't want to disconnect the client, but just ignore the request and continue receiving events.

LukeHankey committed:
Remove whitespace from StatusCodes

kappq commented on server/main.py:
```diff
@@ -44,13 +46,34 @@ async def send(self, data: EventResponse) -> None:
+            await self.send(
+                EventResponse(
+                    type=EventType.ERROR,
+                    data=ErrorData(message="Data not found"),
```
The period is missing at the end of the sentence (just for consistency).

LukeHankey committed:
fix grammar

kappq commented on server/main.py:
```diff
@@ -44,13 +46,34 @@ async def send(self, data: EventResponse) -> None:
+                ),
+            )
+            return
+        except (KeyError, ValidationError, JSONDecodeError):
```
I think that the `JSONDecodeError` should have a status code of `INVALID_REQUEST_DATA`.

It could be just moved up with `TypeError`:
```py
except (TypeError, JSONDecodeError):
    ...
```

LukeHankey committed:
kappa changes

LukeHankey committed:
fix event being None if an error occurs

Only disconnect if event is None, raised by either a WebsocketDisconnect or a RuntimeError
Return a valid EventRequest object in any other error case with default values which won't modify any code
